sklearn/cluster/_optics.py

```diff
@@ -285,7 +287,12 @@ def fit(self, X, y=None):
             )
             warnings.warn(msg, DataConversionWarning)
 
-        X = self._validate_data(X, dtype=dtype)
+        X = self._validate_data(X, dtype=dtype, accept_sparse="csr")
+        if self.metric == "precomputed" and issparse(X):
+            # Set each diagonal to an explicit value so each point is its own neighbor
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore", category=SparseEfficiencyWarning)
+                X.setdiag(X.diagonal())
```

Comment:
Is it necessary ? Can't we let NearestNeighbors do whatever it needs to handle sparse precomputed matrix ?

In addition I think that the parameter description of ``metric`` should be updated to reflect the fact that it supports precomputed sparse graph, as done in the docstrings of nearest neighbors algorithms.